### PR TITLE
Adds PolicySetupTimeoutSeconds option to CalicoNetwork

### PR DESCRIPTION
**Pull request by aaaaaaaalex**
[Sibling PR to be merged first](https://github.com/projectcalico/calico/pull/8469)

## Description

Adds option to CalicoNetwork to enable a policy-setup timeout in the CNI/Felix.

Felix's new option `EndpointStatusPathPrefix` gets set to it's /var/run/calico dir to enable the Felix-side feature, where-in it will create a directory "status".

CNI should have its timeout configured according to the CalicoNetwork field, and it's `EndpointStatusDir` should be set to Felix's /var/run/calico dir plus "/status".

## For PR author

- [x] Tests for change.
- [x] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Comment by aaaaaaaalex** [comment]
Going to mark this PR as Ready for Review, although there are still two merge tests that are failing, that I'm a little confused about.

File: api/v1/installation_types.go [hunk 1]
```diff
@@ -474,6 +474,17 @@ type CalicoNetworkSpec struct {
 	// Sysctl configures sysctl parameters for tuning plugin
 	// +optional
 	Sysctl []Sysctl `json:"sysctl,omitempty"`
+
+	// PolicySetupTimeoutSeconds delays new pods from going ready
+	// until their policy has been programmed in the dataplane.
+	// The specified delay defines the maximum amount of time
+	// that the Calico CNI will wait for policy to be programmed.
```
Comment (caseydavenport) [inline api/v1/installation_types.go:481]: "the Calico CNI _plugin" or just "Calico"

"CNI plugin" refers to a component, and "Calico" refers to the system as a whole.

File: api/v1/installation_types.go [hunk 2]
```diff
@@ -474,6 +474,17 @@ type CalicoNetworkSpec struct {
 	// Sysctl configures sysctl parameters for tuning plugin
 	// +optional
 	Sysctl []Sysctl `json:"sysctl,omitempty"`
+
+	// PolicySetupTimeoutSeconds delays new pods from going ready
```
Comment (caseydavenport) [inline api/v1/installation_types.go:478]: "going ready" should probably be "running containers" - we actually prevent the code from running altogether until the policy is programmed.

File: pkg/controller/installation/validation.go [hunk 3]
```diff
@@ -349,6 +349,13 @@ func validateCustomResource(instance *operatorv1.Installation) error {
 			}
 
 		}
+
+		if instance.Spec.CalicoNetwork.PolicySetupTimeoutSeconds != nil {
```
Comment (caseydavenport) [inline pkg/controller/installation/validation.go:353]: I think we need validation here to prevent setting this in other invalid scenarios - e.g., when Calico networking isn't being used, or when Calico in eBPF mode is being used.

File: pkg/render/node.go [hunk 4]
```diff
@@ -1716,6 +1727,15 @@ func (c *nodeComponent) nodeEnvVars() []corev1.EnvVar {
 		})
 	}
 
+	if c.cfg.Installation.CalicoNetwork != nil &&
+		c.cfg.Installation.CalicoNetwork.PolicySetupTimeoutSeconds != nil &&
+		*c.cfg.Installation.CalicoNetwork.PolicySetupTimeoutSeconds > 0 {
+		nodeEnv = append(nodeEnv, corev1.EnvVar{
+			Name:  "FELIX_ENDPOINTsTATUSPATHPREFIX",
```
Comment (caseydavenport) [inline pkg/render/node.go:1734]: capitalization typo ENDPOINTsTATUS

File: pkg/render/node.go [hunk 5]
```diff
@@ -642,6 +647,8 @@ func (c *nodeComponent) createCalicoPluginConfig() map[string]interface{} {
 		"policy": map[string]interface{}{
 			"type": "k8s",
 		},
+		"policy_setup_timeout_seconds": policySetupTimeoutSeconds,
+		"endpoint_status_dir":          filepath.Join(c.varRunCalicoVolume().VolumeSource.HostPath.Path, "status"),
```
Comment (caseydavenport) [inline pkg/render/node.go:651]: `/var/run/calico/status` is a pretty generic path. Can we make this more intuitive?

e.g., 

`/var/run/calico/policy_program_status` ?

File: pkg/render/node_test.go [hunk 6]
```diff
@@ -199,6 +199,8 @@ var _ = Describe("Node rendering tests", func() {
   "plugins": [
     {
       "type": "calico",
+	  "policy_setup_timeout_seconds": 0,
```
Comment (caseydavenport) [inline pkg/render/node_test.go:202]: Indentation is wrong here and doesn't match the other fields

File: pkg/controller/installation/core_controller.go [hunk 7]
```diff
@@ -657,6 +657,14 @@ func fillDefaults(instance *operator.Installation) error {
 		}
 	}
 
+	// Only fill default policy setup timeout if CalicoCNI is used.
```
Comment (caseydavenport) [inline pkg/controller/installation/core_controller.go:660]: nit, but "CalicoCNI" is neither a valid enum option in our API nor a commonly used representation of "Calico CNI plugin", so saying "if the Calico CNI plugin is used" would be more clear

Comment (caseydavenport) [reply pkg/controller/installation/core_controller.go:660]: Similarly, this comment is self-explanatory by reading the code (or rather, it would be if the code didn't actually do something different from what the comment says!) and doesn't really add any additional context. Would be better if the comment explained _why_ we only default in this case.

> (or rather, it would be if the code didn't actually do something different from what the comment says!) 

We also are checking the LinuxDataplane here, so the comment doesn't match the code :)

File: pkg/controller/installation/validation.go (hunk 3 above)
Comment (caseydavenport) [reply pkg/controller/installation/validation.go:353]: What about Windows? Do we prevent this option in a cluster with Windows nodes? Do we just ignore it?

**Review by caseydavenport — CHANGES_REQUESTED** [review]
Looks pretty good, left a few comments

File: pkg/render/node.go (hunk 5 above)
Comment (caseydavenport) [reply pkg/render/node.go:651]: Would be nice to align with the field name too - we seem to be using `endpoint_status` and `endpointStatus` elsewhere, so maybe this should be `/var/run/calico/endpoint_status`

File: pkg/render/node_test.go [hunk 8]
```diff
@@ -199,6 +199,8 @@ var _ = Describe("Node rendering tests", func() {
   "plugins": [
```
Comment (caseydavenport) [inline pkg/render/node_test.go:199]: I think we need some tests that check non-zero values

File: api/v1/installation_types.go (hunk 2 above)
Comment (aaaaaaaalex) [reply api/v1/installation_types.go:478]: How about "[...] from entering the Running phase" ?

File: pkg/controller/installation/validation.go (hunk 3 above)
Comment (aaaaaaaalex) [reply pkg/controller/installation/validation.go:353]: I've been working off the assumption that we would ignore this option on Windows, how does the operator deal with the issue of windows compat? Does it know if a windows node exists in the cluster?

File: pkg/render/node_test.go (hunk 6 above)
Comment (aaaaaaaalex) [reply pkg/render/node_test.go:202]: VSCode acting up, thanks

Comment (aaaaaaaalex) [reply pkg/render/node_test.go:202]: Found a few other occurrences of tabs instead of spaces in this file's raw json strings, so fixed those up too.

File: pkg/controller/installation/core_controller.go (hunk 7 above)
Comment (aaaaaaaalex) [reply pkg/controller/installation/core_controller.go:660]: Thanks - I have a personal goal to properly re-read all my comments after I amend the code beneath them :rofl:

Comment (aaaaaaaalex) [reply pkg/controller/installation/core_controller.go:660]: Considering removing the comment altogether, since as you say, it's redundant really, and the _why_ really only amounts to "because we haven't covered the other cases yet" :sweat_smile:

File: pkg/render/node.go (hunk 5 above)
Comment (aaaaaaaalex) [reply pkg/render/node.go:651]: Happy to take this onboard, Shaun also suggested:

    I might use a - instead of _ in a dir name (unless there was precedent in an adjacent dir

If that works for you?

File: pkg/controller/installation/validation.go (hunk 3 above)
Comment (aaaaaaaalex) [reply pkg/controller/installation/validation.go:353]: I suppose I could accept it in the operator, ignore it in felix/cni, if thats a worthy paradigm?

File: pkg/render/node_test.go (hunk 8 above)
Comment (aaaaaaaalex) [reply pkg/render/node_test.go:199]: Added some values to validation_test.go now.
Also checking rendering of custom values in node_test.go

File: api/v1/installation_types.go (hunk 2 above)
Comment (caseydavenport) [reply api/v1/installation_types.go]: I think "entering the running phase" is too vague - it doesn't drive the point home that the users application code won't even attempt to be launched. 

"from running" is OK, but IMO still not precise - the pod itself will be running (as in, the pod will have established a network namespace with an  ID and an IP address) - what won't be running are the containers within the pod.

File: pkg/controller/installation/validation.go (hunk 3 above)
Comment (caseydavenport) [reply pkg/controller/installation/validation.go:353]: Typically, we would expect that if a config option isn't supported in a certain configuration, that we return an error and prevent installation of the code in that configuration. We should probably do that for this if Windows is enabled in the configuration.

File: pkg/render/node.go (hunk 5 above)
Comment (caseydavenport) [reply pkg/render/node.go]: :shrug: should be fine either way, main thing is the name. I'd double check the precedent thing though - what other multi-part dir names do we create?

File: pkg/controller/installation/validation.go [hunk 9]
```diff
@@ -355,6 +355,9 @@ func validateCustomResource(instance *operatorv1.Installation) error {
 			if *instance.Spec.CalicoNetwork.PolicySetupTimeoutSeconds < 0 {
 				return fmt.Errorf("Installation CNI spec.PolicySetupTimeoutSeconds negative value is not valid")
 			}
+			if instance.Spec.CalicoNetwork.LinuxDataplane != nil && *instance.Spec.CalicoNetwork.LinuxDataplane != operatorv1.LinuxDataplaneIptables {
```
Comment (caseydavenport) [inline pkg/controller/installation/validation.go:358]: I think we probably want this?

```go
if instance.Spec.CalicoNetwork.LinuxDataplane == nil || *instance.Spec.CalicoNetwork.LinuxDataplane != operatorv1.LinuxDataplaneIptables {
```

Basically if the data plane is unset, then we don't know if it's OK to enable this feature or not and we need to wait until it is.

File: pkg/controller/installation/validation_test.go [hunk 10]
```diff
@@ -791,6 +791,31 @@ var _ = Describe("Installation validation tests", func() {
...
 	})
 
+	Describe("validate CalicoNetwork PolicySetupTimeoutSeconds", func() {
+		It("should return an error when PolicySetupTimeoutSeconds is negative", func() {
+			negative := int32(-1)
+			instance.Spec.CalicoNetwork.PolicySetupTimeoutSeconds = &negative
+			err := validateCustomResource(instance)
+			Expect(err).To(MatchError("Installation CNI spec.PolicySetupTimeoutSeconds negative value is not valid"))
+		})
+		It("should return an error when PolicySetupTimeoutSeconds is set for a non-IPTables dataplane", func() {
```
Comment (caseydavenport) [inline pkg/controller/installation/validation_test.go:801]: nit, but it's nice to have a newline between `It()` clauses for clarity.

File: pkg/controller/installation/validation.go (hunk 9 above)
Comment (caseydavenport) [reply pkg/controller/installation/validation.go:358]: Probably also want a check on WindowsDataplane here

File: pkg/controller/installation/validation.go (hunk 3 above)
Comment (aaaaaaaalex) [reply pkg/controller/installation/validation.go:353]: Im unfamiliar with this part of the operator resources, but my first thought was to check if the windows dataplane config option was non-nil. But, that would presumably only tell me that _some_ nodes may be windows, and I still may want the feature on for non-windows nodes :thinking: any guidance appreciated

File: pkg/render/node.go (hunk 5 above)
Comment (aaaaaaaalex) [reply pkg/render/node.go]: Had a quick skim of some reference docs and occurrences of `Mkdir` in our repo and most I saw seemed to use dashes, so I'll go with that.

File: pkg/controller/installation/validation_test.go (hunk 10 above)
Comment (aaaaaaaalex) [reply pkg/controller/installation/validation_test.go:809]: Fixed

File: pkg/controller/installation/validation.go (hunk 9 above)
Comment (aaaaaaaalex) [reply pkg/controller/installation/validation.go]: Similar to the [question](https://github.com/tigera/operator/pull/3186#discussion_r1508020370) I have above, I think this feature would still be desirable in a mixed cluster with windows and linux, if its possible to continue when that is the case. Maybe I could error if _only_ a Windows dataplane is defined, and no Linux one?

File: api/v1/installation_types.go (hunk 2 above)
Comment (aaaaaaaalex) [reply api/v1/installation_types.go]: Updated to "running containers" now

File: pkg/controller/installation/validation.go (hunk 3 above)
Comment (caseydavenport) [reply pkg/controller/installation/validation.go:353]: There's no such thing as a "Windows only" cluster - they all have at least one Linux node since the control plane components don't run on Windows.

The usual ideal we strive for is that if configuration is accepted it means it is going to be implemented fully. We don't really have a mechanism to report back "This configuration is sort of working but not entirely"

I think your options are:
- Reject the config outright if Windows is enabled.
- Add support for this feature to Windows.

File: pkg/controller/installation/validation.go (hunk 9 above)
Comment (caseydavenport) [reply pkg/controller/installation/validation.go]: > Maybe I could error if only a Windows dataplane is defined, and no Linux one?

Is there a reason we haven't implemented this for Windows?

File: pkg/render/node.go [hunk 11]
```diff
@@ -628,6 +629,10 @@ func (c *nodeComponent) createCalicoPluginConfig() map[string]interface{} {
 
 	apiRoot := c.cfg.K8sServiceEp.CNIAPIRoot()
 
+	LinuxpolicySetupTimeoutSeconds := int32(0)
```
Comment (caseydavenport) [inline pkg/render/node.go:632]: This should be a private variable, and `Policy` should be capitalized

`linuxPolicySetupTimeoutSeconds`

**Review by caseydavenport — APPROVED** [review]

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

**Review by caseydavenport — APPROVED** [review]
